pkg/server/server.go

```diff
@@ -47,20 +48,20 @@ type Server interface {
 // The callbacks are invoked synchronously.
 type Callbacks interface {
 	// OnStreamOpen is called once an xDS stream is open with a stream ID and the type URL (or "" for ADS).
-	OnStreamOpen(int64, string)
+	OnStreamOpen(context.Context, int64, string) error
 	// OnStreamClosed is called immediately prior to closing an xDS stream with a stream ID.
 	OnStreamClosed(int64)
 	// OnStreamRequest is called once a request is received on a stream.
 	OnStreamRequest(int64, *v2.DiscoveryRequest)
 	// OnStreamResponse is called immediately prior to sending a response on a stream.
 	OnStreamResponse(int64, *v2.DiscoveryRequest, *v2.DiscoveryResponse)
 	// OnFetchRequest is called for each Fetch request
-	OnFetchRequest(*v2.DiscoveryRequest)
+	OnFetchRequest(context.Context, *v2.DiscoveryRequest) error
```

Comment:
It's a little awkward that two of these functions now return errors, but the rest do not. In particular, if `OnFetchRequest` can return errors, `OnStreamRequest` is likely to encounter the same errors. Does it make sense to convert all of these to return errors?

Reply:
I was thinking about this just as a way to do auth/pre-processing for a request.  Would it make more sense to add a new Fetch callback like `OnFetchBegin` to separate it and keep the semantics more similar between stream/fetch handling?

I believe any other errors could already be handled by the cache Fetch or Watcher functions.  If there's a good use case for each of the callbacks to return errors, I can add them though.

Reply:
ah that makes sense. I think it's fine as-is, but it's worth documenting why these two functions return an error (eg document explicitly that these would be where you'd stick auth or connection pre-processing).

A separate callback might also make sense, but leaving it this way allows for different behaviors on fetch vs stream open, which might better for some implementations.

Thanks for the quick response!